src/hooks/useTokenList/validateTokenList.ts

```diff
@@ -1,40 +1,34 @@
 import type { TokenInfo, TokenList } from '@uniswap/token-lists'
-import type { Ajv, ValidateFunction } from 'ajv'
+import type { ValidateFunction } from 'ajv'
 
 enum ValidationSchema {
   LIST = 'list',
   TOKENS = 'tokens',
 }
 
-const validator = new Promise<Ajv>(async (resolve) => {
-  const [ajv, schema] = await Promise.all([import('ajv'), import('@uniswap/token-lists/src/tokenlist.schema.json')])
-  const validator = new ajv.default({ allErrors: true })
-    .addSchema(schema, ValidationSchema.LIST)
-    // Adds a meta scheme of Pick<TokenList, 'tokens'>
-    .addSchema(
-      {
-        ...schema,
-        $id: schema.$id + '#tokens',
-        required: ['tokens'],
-      },
-      ValidationSchema.TOKENS
-    )
-  resolve(validator)
-})
-
 function getValidationErrors(validate: ValidateFunction | undefined): string {
   return (
-    validate?.errors?.map((error) => [error.dataPath, error.message].filter(Boolean).join(' ')).join('; ') ??
+    validate?.errors?.map((error) => [error.instancePath, error.message].filter(Boolean).join(' ')).join('; ') ??
     'unknown error'
   )
 }
 
+async function loadValidator(schema: ValidationSchema): Promise<ValidateFunction> {
+  const [, validatorModule] = await Promise.all([
+    import('ajv'),
+    schema === ValidationSchema.LIST
+      ? import('__generated__/validateTokenList')
+      : import('__generated__/validateTokens'),
+  ])
+  return (await validatorModule.default) as ValidateFunction
+}
+
 /**
  * Validates an array of tokens.
  * @param json the TokenInfo[] to validate
  */
 export async function validateTokens(json: TokenInfo[]): Promise<TokenInfo[]> {
-  const validate = (await validator).getSchema(ValidationSchema.TOKENS)
+  const validate = await loadValidator(ValidationSchema.TOKENS)
```

Comment:
Would you consider a single `validate(schema: ValidationSchema, data: unknown)` function?

Reply:
i changed it, but i can't say i think it's much better - you lose type specificity and it doesn't save much in terms of LOC